src/common/hist_builder.h

```diff
@@ -0,0 +1,156 @@
+                                       rows[i + Prefetch::kPrefetchOffset] * n_features;
+      const size_t icol_end_prefetch = any_missing ?  row_ptr[rows[i+Prefetch::kPrefetchOffset]+1] :
+                                      icol_start_prefetch + n_features;
+
+      PREFETCH_READ_T0(pgh + two * rows[i + Prefetch::kPrefetchOffset]);
+      for (size_t j = icol_start_prefetch; j < icol_end_prefetch;
+        j+=Prefetch::GetPrefetchStep<uint32_t>()) {
+        PREFETCH_READ_T0(gradient_index + j);
+      }
+    } else if (is_root) {
+      nodes_ids[ri] = 0;
+    }
+
+    const BinIdxType* gr_index_local = gradient_index + icol_start;
+    FPType* hist_data = hists[nid].data();
+
+    for (size_t j = 0; j < row_size; ++j) {
+      const uint32_t idx_bin = two * (static_cast<uint32_t>(gr_index_local[j]) + (
+                                      any_missing ? 0 : offsets[j]));
+      hist_data[idx_bin]   += pgh[idx_gh];
+      hist_data[idx_bin+1] += pgh[idx_gh+1];
+    }
+  }
+}
+
+/*!
+ * \brief builder for histograms of gradient statistics
+ */
+template<typename GradientSumT>
+class GHistBuilder {
+ public:
+  using GHistRowT = GHistRow<GradientSumT>;
+  GHistBuilder() = default;
+  GHistBuilder(size_t nthread, uint32_t nbins) : nthread_{nthread}, nbins_{nbins} {}
+
+  // construct a histogram via histogram aggregation
+  template <typename BinIdxType, bool any_missing, bool is_root>
+  void BuildHist(const std::vector<GradientPair>& gpair,
+                 const uint32_t* rows,
+                 const uint32_t row_begin,
+                 const uint32_t row_end,
+                 const GHistIndexMatrix& gmat,
+                 const BinIdxType* numa_data,
+                 uint16_t* nodes_ids,
+                 std::vector<std::vector<GradientSumT>>* p_hists,
+                 const uint16_t* mapping_ids) {
+    const size_t nrows = row_end - row_begin;
+    const size_t no_prefetch_size = Prefetch::NoPrefetchSize(nrows);
+
+    if (is_root) {
+        // contiguous memory access, built-in HW prefetching is enough
+        BuildHistKernel<GradientSumT, false, BinIdxType, true, any_missing>(
+          gpair, rows, row_begin, row_end, gmat, numa_data, nodes_ids, p_hists, mapping_ids);
+    } else {
+        BuildHistKernel<GradientSumT, true, BinIdxType, false, any_missing>(
+          gpair, rows, row_begin, row_end - no_prefetch_size,
+          gmat, numa_data, nodes_ids, p_hists, mapping_ids);
+        BuildHistKernel<GradientSumT, false, BinIdxType, false, any_missing>(
+          gpair, rows,  row_end - no_prefetch_size, row_end,
+          gmat, numa_data, nodes_ids, p_hists, mapping_ids);
+    }
+  }
+
+  uint32_t GetNumBins() const {
+      return nbins_;
+  }
+
+ private:
+  /*! \brief number of threads for parallel computation */
+  size_t nthread_ { 0 };
+  /*! \brief number of all bins over all features */
+  uint32_t nbins_ { 0 };
```

Comment:
This also doesn't seem to be used within the class?

Reply:
deleted. Thanks!